Mathlib/Data/UInt.lean

```diff
@@ -1,10 +1,231 @@
+import Mathlib.Data.Nat.Basic
+import Mathlib.Data.Fin.Basic
+import Mathlib.Algebra.Group.Defs
+import Mathlib.Algebra.GroupWithZero.Defs
 
-theorem Fin.val_eq_of_lt : a < n.succ → (@Fin.ofNat n a).val = a := Nat.mod_eq_of_lt
+lemma UInt8.size_positive : 0 < UInt8.size := by decide
 
-theorem UInt32.val_eq_of_lt : a < UInt32.size → (UInt32.ofNat a).val = a := Fin.val_eq_of_lt
+lemma UInt16.size_positive : 0 < UInt16.size := by decide
+
+lemma UInt32.size_positive : 0 < UInt32.size := by decide
+
+lemma UInt64.size_positive : 0 < UInt64.size := by decide
+
+lemma USize.size_positive : 0 < USize.size := by
+  simp only [USize.size]
+  cases System.Platform.numBits_eq with
+  | inl eq32 => simp [eq32]
+  | inr eq64 => simp [eq64]
+
+set_option hygiene false
+macro "genIntDeclars" typeName:ident : command => do
+  `(
+    namespace $typeName
+      instance : Zero (Fin size) := (Nat.succ_pred_eq_of_pos size_positive) ▸ inferInstance
```

Comment:
I think I missed this earlier.
```suggestion
      instance : Zero (Fin size) := inferInstanceAs (Zero (Fin (_ + 1)))
```
The `▸` must be avoided in instances, because it generates an `eq.rec` application.  And in general, this `eq.rec` will not reduce definitionally, unless the types are definitionally equal.  But if the types are definitionally equal, then it's better to make that explicit by writing `inferInstanceAs`.

Reply:
@gebner 

Thanks, I wasn't sure whether this would cause issues down the road, but I'll remove the `▸ ` uses. inferInstanceAs throws an error for USize.size:

```
instance : Zero (Fin USize.size) := inferInstanceAs (Zero (Fin (_ + 1)))

/-
type mismatch
  inferInstanceAs (Zero (Fin (?m.839 + 1)))
has type
  Zero (Fin (?m.839 + 1)) : Type
but is expected to have type
  Zero (Fin USize.size) : Type
-/
```

I'll inline all of the typeclass members that I can and protect the ones that are more tricky. In Lean 3, I sometimes found it difficult to figure out where a particular lemma was for a given type because instead of being `Fin.lemma` it was a member of some typeclass I wasn't able to guess, and instances don't always have guessable names. I understand why it's favorable outside of that to inline, so is the guidance that they should use librarySearch or find to look for lemmas?

Reply:
Oh, USize is a problem indeed because `USize` does not definitionally reduce to `Nat.succ _`....

Now I recall what we do in mathlib sometimes, instead of `Inhabited (Fin n)` we write `Zero (Fin n)` but it's the same idea.  Maybe we should really go for this approach.

> I understand why it's favorable outside of that to inline, so is the guidance that they should use librarySearch or find to look for lemmas?

Yes, library search is always a good idea.  We're never going to duplicate all lemmas we have for groups (nor should we), so you're always going to have to look for lemmas about type classes.  And if you don't look beyond Fin, then you're missing out on great lemmas.

I agree that we should have better tooling for this.  Questions like: what type classes does a type implement? what lemmas are there about objects of a given type?  are hard to figure out at the moment.  Ideally those would be available in the documentation (and via commands).

Reply:
I'm not sure how to get the `Zero (Fin n)` route to work without duplicating the basic implementations in the Fin module for both `Fin (n+1)` and an `[Inhabited (Fin n)]`; the `Zero` typeclass doesn't say anything about the element it produces, so even though `AddMonoid` inherits `Zero`, you need an actual instance of `Zero` to be able to prove e.g. `zero_add`, you just end up with a goal of `(a : Fin n), a + (?m.4307 : Fin n) = a`, where the metavariable is the `Zero.zero`.

The `[Inhabited (Fin n)]` route does add the extra typeclass bound, but `instance : Inhabited (Fin (no_index (n+1)))` already exists in core, so even if you add the Inhabited bound, you get all of the `Fin n.succ` instances for free. 

The fin module in mathlib3 seems to implement has_zero `fin n+1`, and I couldn't find any alternative uses of `has_zero` in the fin module. My copy of mathlib 3 is oldish, but this doesn't seem to work:

```
example (n : nat) [inhabited (fin n)] (a b : fin n) : a + b = b + a := 
begin
  library_search,
end
```

Reply:
The `has_zero (fin n)` assumption is used in other parts of mathlib (namely Euclidean half spaces as a manifold with boundaries).  Okay, it's a somewhat obscure and otherwise unused part of mathlib but at least this wouldn't be the first time we use this approach.

> the Zero typeclass doesn't say anything about the element it produces

Ah I see.  We want to define `(0 : Fin n)` so that `(0 : Fin n).val = 0` is true definitionally.  And so it doesn't matter whether we assume `Inhabited` or `Zero` or even `Nonempty`, since we only need that type class instance to conclude `0 < n`.  Then the `Inhabited` assumption seems like the better idea.

Reply:
I pushed the relevant changes in adaf634, inlining where possible which removed the bad `@[simp]` attributes. USize is now implemented as well and the macro is marked local, so I'll mark those issues as resolved.